### ISSUE TITLE
How worried should we be about transmuting `Vec`?

### ISSUE THREAD

**Issue by Enet4** [nabijaczleweli/safe-transmute-rs#16]
Although our guarded vector conversions seem to work in practice, we seem to be disregarding one of the safety conditions stated in the [docs](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.from_raw_parts):

> - `ptr` needs to have been previously allocated via `String`/`Vec<T>` (at least, it's highly likely to be incorrect if it wasn't).
> - `ptr`'s `T` **needs to have the same size and alignment as it was allocated with**.
> - `length` needs to be less than or equal to `capacity`.
> - `capacity` needs to be the capacity that the pointer was allocated with.

This is a pity, because one of my use cases involves converting byte arrays into other third-party array types of different sizes without memory reallocation. Our tests do not seem to be calling this a problem, but I wonder whether this would misbehave at certain edge cases, or even whether this could indeed change from one compiler version to another.

**Comment by nabijaczleweli** [comment]
How worried we are depends on what can happen. Considering we convert from `Vec<u8>` to `Vec<T>` there aren't that many cases:

Assuming `u = sizeof(u8) && u = alignof(u8)`.
Assuming `t = alignof(T) && sizeof(T) <= t`.

1. `sizeof(T) == alignof(T) == u` -- perfectly in order and actually valid

2. `sizeof(T) == alignof(T) && t = ku` where `k` in N -- I'd say we risk misaligned access, which some CPUs trap on. Consider required 4-byte alignment on 4-byte accesses and no alignment on 1-byte accesses, you could have `Vec<u8>::data() % 4 != 0 => Vec<u32>::data() % 4 != 0 => Vec<u32>::at(_) traps`. But I'm not an expert in how that works (and you seem to be).

3. `sizeof(T) != alignof(T) && t = ku` where `k` in N -- Same as above and also we're trashing some bytes. Now, they shouldn't matter, but who the fuck knows.

That should about cover it, right? I can hardly think of a different case where doing that would be invalid (I mean it's probably UB either way but I'm no Rust language lawyer).

**Comment by minno726** [comment]
I think this is still undefined behavior. For a concrete example, [std's implementation of dealloc on Windows](https://github.com/rust-lang/rust/blob/190feb65290d39d7ab6d44e994bd99188d339f16/src/libstd/sys/windows/alloc.rs#L46-L57) checks the alignment of the type, and behaves differently depending on that alignment. If you call `transmute_to_bytes_vec` with a type that has a high alignment, it can cause memory corruption.

Code:

```
use std::mem::{size_of, forget};

// Copied from safe_transmute::to_bytes 0.10.1
fn transmute_to_bytes_vec<T>(mut from: Vec<T>) -> Vec<u8> {
    unsafe {
        let capacity = from.capacity() * size_of::<T>();
        let len = from.len() * size_of::<T>();
        let ptr = from.as_mut_ptr();
        forget(from);
        Vec::from_raw_parts(ptr as *mut u8, len, capacity)
    }
}

#[repr(C)]
#[repr(align(32))]
#[derive(Copy, Clone)]
struct Test {
    value: [u8; 32],
}

fn main() {
    let values = vec![Test { value: [42; 32] }; 1000];
    let bytes = transmute_to_bytes_vec(values);
}
```

Result:

```
D:\Users\will\code\misc\Rust\example>rustc --version
rustc 1.33.0 (2aa4c46cf 2019-02-28)

D:\Users\will\code\misc\Rust\example>cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `target\debug\example.exe`
error: process didn't exit successfully: `target\debug\example.exe` (exit code: 0xc0000374, STATUS_HEAP_CORRUPTION)
```

**Comment by Enet4** [comment]
Thank you for reporting this issue, @minno726. We'll be sure to keep this in mind for version 0.11.0.

**Comment by Enet4** [comment]
With #49 now merged, our conditions of transmuting for vectors have been restricted even more.

> - `ptr` needs to have been previously allocated via `String`/`Vec<T>` (at least, it's highly likely to be incorrect if it wasn't).

This was already ensured by the functions' main argument.

> - `ptr`'s `T` **needs to have the same size and alignment as it was allocated with**.

We now require the `size_of` and `align_of` of the source to match the target's, returning an `IncompatibleVecTarget` error if any of the two are different.

> - `length` needs to be less than or equal to `capacity`.

Still check.

> - `capacity` needs to be the capacity that the pointer was allocated with.

Considering that the size and alignment have to be the same, the length and capacity of the new vector will also match. The `transmute_vec` function in the `base` module does not check for compatible `size_of`/`align_of`, but will still recalculate the length and capacity to match the number of allocated bytes without overrunning the buffer. This one has been documented and marked `unsafe`. Its use is still pretty dangerous, and not recommended. Check.

We have also integrated `miri` with CI, which is known to check for inconsistent deallocations such as the one reported here.

I think we're OK with closing this issue now (as always, do leave some feedback if I overlooked something). Thank you to all who intervened in this major concern.
